splink/cluster_metrics.py

```diff
@@ -1,53 +1,123 @@
+from typing import Dict, List
+
 from splink.splink_dataframe import SplinkDataFrame
 
 
-def _size_density_sql(
+def _node_degree_sql(
     df_predict: SplinkDataFrame,
     df_clustered: SplinkDataFrame,
-    threshold_match_probability: float,
     composite_uid_edges_l: str,
+    composite_uid_edges_r: str,
     composite_uid_clusters: str,
-):
-    """Generates sql for computing cluster size and density at a given threshold.
+    threshold_match_probability: float,
+) -> List[Dict[str, str]]:
+    """
+    Generates sql for computing node degree per node, at a given edge threshold.
+
+    This is includes nodes with no edges, as identified via the clusters table.
 
     Args:
         df_predict (SplinkDataFrame): The results of `linker.predict()`.
         df_clustered (SplinkDataFrame): The outputs of
                 `linker.cluster_pairwise_predictions_at_threshold()`.
+        composite_uid_edges_l (str): unique id for left-hand edges.
+        composite_uid_edges_r (str): unique id for right-hand edges.
+        composite_uid_clusters (str): unique id for clusters.
         threshold_match_probability (float): Filter the pairwise match
             predictions to include only pairwise comparisons with a
             match_probability above this threshold.
-        composite_uid_edges_l (str): unique id for left-hand edges.
-        composite_uid_clusters (str): unique id for clusters.
-
 
     Returns:
-        sql string for computing cluster size and density
+        array of dicts, with sql string and output name
+        for computing cluster size and density
     """
-
     sqls = []
-    # Count edges per node at or above a given match probability
+    edges_tn = df_predict.physical_name
+    clusters_tn = df_clustered.physical_name
+
     sql = f"""
         SELECT
-            {composite_uid_edges_l} AS edge_group_id,
-            COUNT(*) AS count_edges
-        FROM {df_predict.physical_name}
-        WHERE match_probability >= {threshold_match_probability}
-        GROUP BY {composite_uid_edges_l}
+            *
+        FROM
+            {edges_tn}
+        WHERE
+            match_probability >= {threshold_match_probability}
     """
-    sql = {"sql": sql, "output_table_name": "__splink__edges_per_node"}
-    sqls.append(sql)
+    output_tn = "__splink__truncated_edges"
+    sql_info = {"sql": sql, "output_table_name": output_tn}
+    sqls.append(sql_info)
+
+    sql = f"""
+        SELECT
+            {composite_uid_edges_l} AS node,
+            {composite_uid_edges_r} AS neighbour
+        FROM {output_tn}
+            UNION ALL
+        SELECT
+            {composite_uid_edges_r} AS node,
+            {composite_uid_edges_l} AS neighbour
+        FROM {output_tn}
+    """
+    output_tn = "__splink__all_nodes"
+    sql_info = {"sql": sql, "output_table_name": output_tn}
+    sqls.append(sql_info)
```

Comment:
Same here for `output_tn`, but again, personal preference, so feel free to ignore!

Reply:
I have renamed on a case-by-case basis with hopefully clear names. I'm not sure if you're suggesting ditching this variable entirely - the reason I have it is a) single point for any name-changing and b) so I can click on variables on [e.g. this line](https://github.com/moj-analytical-services/splink/blob/5a1d1864e3a03b0ae45d6d51bf81a523db794e33/splink/cluster_metrics.py#L71) and it will take me up to the definition (which has the corresponding SQL next to it).

Appreciate that this function is pretty simple currently though, so neither of these points are particularly important in this case, so happy if you still think having the literal names aids clarity

Reply:
Names all look good to me and happy to keep the variable